audio/src/range_set.rs

```diff
@@ -1,124 +1,138 @@
-use std::cmp::{max, min};
+#![allow(clippy::suspicious_operation_groupings)]
+
 use std::fmt;
-use std::slice::Iter;
 
-#[derive(Copy, Clone, Debug)]
-pub struct Range {
-    pub start: usize,
-    pub length: usize,
+mod range;
+pub(crate) use range::Range;
+
+/// An ordered set of ranges.
+///
+/// ### Invariants:
+/// 1. `ranges` is ordered by `range.start()`. This *must* be  upheld for the implementation to be
+/// correct.
+/// 2. The ranges do not touch. Upon insertion of a range that touches another, instead, those
+/// ranges are merged. This *must* be  upheld for the implementation to be correct.
+/// 3. None of the ranges are empty. Adding empty ranges causes needless performance degradation.
```

Comment:
Please don't differentiate between different "strengths" of invariants. Just enforce that they are always uphold for code outside of this module.

If these invariants do not affect the use of the public api, they should not be documented as rustdoc but as normal comments inside of the struct.

> None of the ranges are empty.

Strictly speaking, this is a direct consequency of the 2. invariant: An empty range touches itself. However, this might be not directly obvious, so it's good to have it documented.

> Adding empty ranges causes needless performance degradation.

There's no need to worry much about performance. I hardly believe it would affect librespot in a measurable way.  If you still want to consider it, you should set up benchmarks for different use cases. 

(One possibility that _could_ be faster: A `BTreeMap` with "start" as key and "length" as value. Many operations would be O(log n), although there's no way to merge two maps in the most efficient way without accessing its internals. So probably it's not really worthwhile.)

Reply:
> Please don't differentiate between different "strengths" of invariants.

Gone.

> An empty range touches itself.

If we define `touching` of two sets S1 and S2 as stating that there exists an element b in the boundary 𝛿S1 of S1 such that b ∈ closure(S2), then since the boundary 𝛿Ø of the empty set Ø is again Ø, no set touches the empty set, including the empty set itself.  Though I admit this is a bit theoretical to be immediately obvious.

> There's no need to worry much about performance.

That is fair, I will remove the comment

Reply:
Well, it's really a matter of definition. First of all, you don't define of what kind of sets you're speaking of. Sets of natural numbers don't make any sense, because on discrete topologies, boundary and closure of a set are the set itself. So you're speaking of half-open sets on real numbers?

But whatever theoretical definition you choose, here's the implemention of `touches` in the code:
```rust
self.start() <= other.end() && self.end() >= other.start()
```

The empty range touches itself.

Reply:
You are right. But should we really leave the third invariant implicit?

Reply:
> But should we really leave the third invariant implicit?

No, of course not :)